akamai/cache-client.go

```diff
@@ -0,0 +1,243 @@
+// APIs, their docs (https://developer.akamai.com/introduction/Client_Auth.html)
+// provide a  description of the required generation process.
+func (cpc *CachePurgeClient) constructAuthHeader(request *http.Request, body []byte, apiPath string, nonce string) (string, error) {
+	// The akamai API is very time sensitive (recommending reliance on a stratum 2
+	// or better time source) and, although it doesn't say it anywhere, really wants
+	// the timestamp to be in the UTC timezone for some reason.
+	timestamp := cpc.clk.Now().UTC().Format(timestampFormat)
+	header := fmt.Sprintf(
+		"EG1-HMAC-SHA256 client_token=%s;access_token=%s;timestamp=%s;nonce=%s;",
+		cpc.clientToken,
+		cpc.accessToken,
+		timestamp,
+		nonce,
+	)
+	bodyHash := sha256.Sum256(body)
+	tbs := fmt.Sprintf(
+		"%s\t%s\t%s\t%s\t%s\t%s\t%s",
+		"POST",
+		cpc.apiScheme,
+		cpc.apiHost,
+		apiPath,
+		"", // We don't need to send any signed headers for a purge so this can be blank
+		base64.StdEncoding.EncodeToString(bodyHash[:]),
+		header,
+	)
+
+	// Create signing key using a HMAC of the client secret over the timestamp
+	h := hmac.New(sha256.New, []byte(cpc.clientSecret))
+	h.Write([]byte(timestamp))
+	key := make([]byte, base64.StdEncoding.EncodedLen(32))
+	base64.StdEncoding.Encode(key, h.Sum(nil))
+
+	h = hmac.New(sha256.New, key)
+	h.Write([]byte(tbs))
+	return fmt.Sprintf(
+		"%ssignature=%s",
+		header,
+		base64.StdEncoding.EncodeToString(h.Sum(nil)),
+	), nil
+}
+
+// purge actually sends the individual requests to the Akamai endpoint and checks
+// if they are successful
+func (cpc *CachePurgeClient) purge(urls []string) error {
+	purgeReq := purgeRequest{
+		Objects: urls,
+		Action:  "remove",
+		Type:    "arl",
+		// Domain:  "staging",
+	}
+	reqJSON, err := json.Marshal(purgeReq)
+	if err != nil {
+		return errFatal(err.Error())
+	}
+
+	req, err := http.NewRequest(
+		"POST",
+		fmt.Sprintf("%s%s", cpc.apiEndpoint, purgePath),
+		bytes.NewBuffer(reqJSON),
+	)
+	if err != nil {
+		return errFatal(err.Error())
+	}
+
+	// Create authorization header for request
+	authHeader, err := cpc.constructAuthHeader(
+		req,
+		reqJSON,
+		purgePath,
+		core.RandomString(16),
+	)
+	if err != nil {
+		return errFatal(err.Error())
+	}
+	req.Header.Set("Authorization", authHeader)
+	req.Header.Set("Content-Type", "application/json")
+
+	rS := cpc.clk.Now()
+	resp, err := cpc.client.Do(req)
+	cpc.stats.TimingDuration("CCU.PurgeRequestLatency", time.Since(rS), 1.0)
+	if err != nil {
+		return err
+	}
+	if resp.Body == nil {
+		return fmt.Errorf("No response body")
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	// Check purge was successful
+	var purgeInfo purgeResponse
+	err = json.Unmarshal(body, &purgeInfo)
+	if err != nil {
+		return err
+	}
+	if purgeInfo.HTTPStatus != 201 || resp.StatusCode != 201 {
+		return fmt.Errorf("Incorrect HTTP status code")
```

Comment:
I think we should include the error body here too, to help diagnose. From https://api.ccu.akamai.com/ccu/v2/docs/index.html, it looks like it'll be `application/api-problem+json` which should be fine to log.